File: internal/ignore/ignore.go
// Package ignore provides rule-based filtering for excluded paths and symbols
// used by tld analyze and tld check commands.
package ignore

import (
	"path/filepath"
	"strings"

	"github.com/bmatcuk/doublestar/v4"
)

// Rules holds gitignore-style exclusion patterns loaded from the workspace configuration file.
type Rules struct {
	Exclude []string `yaml:"exclude,omitempty"`
}

// Merge combines multiple rule sets into a single exclusion list.
func Merge(rules ...*Rules) *Rules {
	merged := &Rules{}
	seen := make(map[string]struct{})
	for _, ruleSet := range rules {
		if ruleSet == nil {
			continue
		}
		for _, pattern := range ruleSet.Exclude {
			pattern = strings.TrimSpace(pattern)
			if pattern == "" {
				continue
			}
			if _, ok := seen[pattern]; ok {
				continue
			}
			seen[pattern] = struct{}{}
			merged.Exclude = append(merged.Exclude, pattern)
		}
	}
	if len(merged.Exclude) == 0 {
		return nil
	}
	return merged
}

// ShouldIgnorePath returns true if the given file or folder path matches any exclusion pattern.
// The path can be absolute or relative; matching is performed against both the full path and base name.
func (r *Rules) ShouldIgnorePath(path string) bool {
	if r == nil {
		return false
	}
	path = normalizePath(path)
	base := filepath.Base(path)
	for _, pattern := range r.Exclude {
		if pattern == "" {
			continue
		}
		normalizedPattern := normalizePattern(pattern)
		if matchPattern(normalizedPattern, path) || matchPattern(normalizedPattern, base) {
			return true
		}
		if before, ok := strings.CutSuffix(normalizedPattern, "/"); ok {
			trimmed := before
			if path == trimmed || strings.HasPrefix(path, trimmed+"/") || base == trimmed {
				return true
			}
		}
	}
	return false
}

// ShouldIgnoreFile returns true if the given file path is excluded.
func (r *Rules) ShouldIgnoreFile(path string) bool {
	return r.ShouldIgnorePath(path)
}

// ShouldIgnoreFolder returns true if the given folder path is excluded.
func (r *Rules) ShouldIgnoreFolder(path string) bool {
	return r.ShouldIgnorePath(path)
}

// ShouldIgnoreSymbol returns true if the given symbol name matches any exclusion pattern.
func (r *Rules) ShouldIgnoreSymbol(name string) bool {
	if r == nil {
		return false
	}
	name = strings.TrimSpace(name)
	for _, pattern := range r.Exclude {
		if pattern == "" {
			continue
		}
		normalizedPattern := normalizePattern(pattern)
		if matchPattern(normalizedPattern, name) {
			return true
		}
	}
	return false
}

// matchPattern matches a value against a pattern using gitignore-style glob syntax.
// It falls back to exact string equality if the glob is invalid.
func matchPattern(pattern, value string) bool {
	matched, err := doublestar.Match(pattern, value)
	if err != nil {
		return pattern == value
	}
	return matched
}

func normalizePath(path string) string {
	path = strings.TrimSpace(path)
	path = filepath.ToSlash(path)
	path = strings.TrimPrefix(path, "./")
	path = strings.TrimPrefix(path, "/")
	return path
}

func normalizePattern(pattern string) string {
	pattern = strings.TrimSpace(pattern)
	pattern = filepath.ToSlash(pattern)
	pattern = strings.TrimPrefix(pattern, "./")
	return pattern
}

File: src/analyzer/ignore.rs
use glob::Pattern;
#[derive(Debug, Default, Clone)]
pub struct Rules {
    pub exclude: Vec<String>,
}

impl Rules {
    pub fn new(exclude: Vec<String>) -> Self {
        Self { exclude }
    }

    pub fn should_ignore_path(&self, path: &str) -> bool {
        let normalized_path = self.normalize_path(path);
        let path_segments: Vec<&str> = normalized_path.split('/').collect();
        let base_name = path_segments.last().copied().unwrap_or("");

        for pattern in &self.exclude {
            if pattern.is_empty() {
                continue;
            }
            let normalized_pattern = self.normalize_pattern(pattern);
            let trimmed_pattern = normalized_pattern.strip_suffix('/').unwrap_or(&normalized_pattern);
            let is_dir_pattern = normalized_pattern.ends_with('/');

            // 1. Direct match or glob match on the full normalized path
            if self.match_pattern(&normalized_pattern, &normalized_path) {
                return true;
            }

            // 2. Match against any segment of the path (for directory patterns or base names)
            for segment in &path_segments {
                if is_dir_pattern {
                    if *segment == trimmed_pattern {
                        return true;
                    }
                } else if self.match_pattern(&normalized_pattern, segment) {
                    return true;
                }
            }

            // 3. Fallback for the base name itself
            if self.match_pattern(&normalized_pattern, base_name) {
                return true;
            }
        }
        false
    }

    fn match_pattern(&self, pattern: &str, value: &str) -> bool {
        match Pattern::new(pattern) {
            Ok(p) => p.matches(value),
            Err(_) => pattern == value,
        }
    }

    fn normalize_path(&self, path: &str) -> String {
        path.trim()
            .replace('\\', "/")
            .trim_start_matches("./")
            .trim_start_matches('/')
            .to_string()
    }

    fn normalize_pattern(&self, pattern: &str) -> String {
        pattern
            .trim()
            .replace('\\', "/")
            .trim_start_matches("./")
            .to_string()
    }
}
